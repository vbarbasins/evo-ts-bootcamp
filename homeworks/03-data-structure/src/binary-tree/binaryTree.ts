export type TreeNode<T> = {
  value: T;
  left: TreeNode<T> | null;
  right: TreeNode<T> | null;
};

export enum TraverseType {
  DFS_INORDER = 'DFS_INORDER',
  DFS_PREORDER = 'DFS_PREORDER',
  DFS_POSTORDER = 'DFS_POSTORDER',
  BFS = 'BFS',
}

export interface Tree<T> {
  setTreeNode(treeNode: TreeNode<T>): this;
  traverse(type: TraverseType): T[];
  getColumn(column: number): T[];
}

export class BinaryTree<T> implements Tree<T> {
  protected rootNode: TreeNode<T>;

  constructor(treeNode: TreeNode<T>) {
    this.rootNode = treeNode;
  }

  setTreeNode(treeNode: TreeNode<T>): this {
    this.rootNode = treeNode;
    return this;
  }

  getColumn(targetColumn: number): T[] {
    const { value, left, right } = this.rootNode;
    const currentColumn = 0;
    const result: T[] = [];
    if (currentColumn === targetColumn) result.push(value);
    const traverseLeft = left === null ? [] : new BinaryTree(left).getColumn(targetColumn + 1);
    const traverseRight = right === null ? [] : new BinaryTree(right).getColumn(targetColumn - 1);
    return [...traverseLeft, ...result, ...traverseRight];
  }

  traverse(type: TraverseType): T[] {
    const { value, left, right } = this.rootNode;
    const traverseLeft = left === null ? [] : new BinaryTree(left).traverse(type);
    const traverseRight = right === null ? [] : new BinaryTree(right).traverse(type);

    function assertNever(arg: never): never {
      throw new Error(`Unexpected argument: ${arg}`);
    }

    switch (type) {
      case TraverseType.BFS: {
        const result: T[] = [];
        const queue: TreeNode<T>[] = [this.rootNode];
        while (queue.length > 0) {
          const node = queue.shift()!;
          result.push(node.value);
          if (node.left) queue.push(node.left);
          if (node.right) queue.push(node.right);
        }
        return result;
      }
      case TraverseType.DFS_INORDER: return [...traverseLeft, value, ...traverseRight];
      case TraverseType.DFS_PREORDER: return [value, ...traverseLeft, ...traverseRight];
      case TraverseType.DFS_POSTORDER: return [...traverseLeft, ...traverseRight, value];
      default: return assertNever(type);
    }
  }
}
