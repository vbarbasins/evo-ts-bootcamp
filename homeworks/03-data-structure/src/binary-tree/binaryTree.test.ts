import { BinaryTree, TraverseType } from './binaryTree';

/*             43
*             /  \
*          40     46
*         /  \   /  \
*       10    2 1    5
*              \
*               67
*/
const givenTree = new BinaryTree({
  value: 43,
  left: {
    value: 40,
    left: { value: 10, left: null, right: null },
    right: {
      value: 2,
      left: null,
      right: { value: 67, left: null, right: null },
    },
  },
  right: {
    value: 46,
    left: { value: 1, left: null, right: null },
    right: { value: 5, left: null, right: null },
  },
});

describe('BinaryTree', () => {
  describe('setTreeNode', () => {
    it('should set new tree node as root node', () => {
      const tree = new BinaryTree({ value: 1, left: null, right: null });
      tree.setTreeNode({ value: 2, left: null, right: null });
      expect(tree).toMatchSnapshot();
    });
  });

  describe('traverse', () => {
    it('should throw an error with unexpected traverse type', () => {
      const given = new BinaryTree({ value: 1, left: null, right: null });
      expect(() => {
        given.traverse('unexpected type' as TraverseType);
      }).toThrow('Unexpected argument: unexpected type');
    });

    describe.each([
      [TraverseType.DFS_INORDER, givenTree, [10, 40, 2, 67, 43, 1, 46, 5]],
      [TraverseType.DFS_PREORDER, givenTree, [43, 40, 10, 2, 67, 46, 1, 5]],
      [TraverseType.DFS_POSTORDER, givenTree, [10, 67, 2, 40, 1, 5, 46, 43]],
      [TraverseType.BFS, givenTree, [43, 40, 46, 10, 2, 1, 5, 67]],
    ])('.traverse(%p)', (traverseType, given, expected) => {
      it(`should return ${expected}`, () => {
        expect(given.traverse(traverseType)).toStrictEqual(expected);
      });
    });
  });

  describe('getColumn', () => {
    describe.each([
      [0, givenTree, [2, 43, 1]],
      [-1, givenTree, [40]],
      [1, givenTree, [67, 46]],
      [-3, givenTree, []],
    ])('.getColumn(%i)', (column, given, expected) => {
      it(`should return ${expected}`, () => {
        expect(given.getColumn(column)).toStrictEqual(expected);
      });
    });
  });
});
