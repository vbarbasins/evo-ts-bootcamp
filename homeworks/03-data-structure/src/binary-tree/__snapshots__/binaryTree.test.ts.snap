// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BinaryTree setTreeNode should set new tree node as root node 1`] = `
BinaryTree {
  "rootNode": Object {
    "left": null,
    "right": null,
    "value": 2,
  },
}
`;
